compiler/src/main/scala/org/scalajs/core/compiler/PreTyperComponent.scala

```diff
@@ -0,0 +1,121 @@
+ *
+ *  Would become:
+ *  {{{
+ *  class $anon extends ... {
+ *    @WasPublicBeforeTyper val foo = ???
+ *    @WasPublicBeforeTyper var bar = ???
+ *    @WasPublicBeforeTyper def baz = ???
+ *    private val foo2 = ???
+ *    private var bar2 = ???
+ *    private def baz2 = ???
+ *  }
+ *  }}}
+ *
+ *  And after `typer` (if has SI-9487) will be:
+ *  {{{
+ *  class $anon extends ... {
+ *    @WasPublicBeforeTyper private[this] var foo = ???
+ *    private <stable> <accessor> def foo = ??? // Needs fix
+ *
+ *    @WasPublicBeforeTyper private[this] var bar = ???
+ *    private <accessor> def bar = ??? // Needs fix
+ *    private <accessor> def bar_=(...) = ??? // Needs fix
+ *
+ *    @WasPublicBeforeTyper private def baz = ??? // Needs fix
+ *    ...
+ *  }
+ *  }}}
+ *
+ *  @author Nicolas Stucki
+ */
+abstract class PreTyperComponent
+    extends plugins.PluginComponent with transform.Transform with Compat210Component {
+
+  import global._
+
+  val phaseName = "jspretyper"
+
+  override protected def newTransformer(unit: CompilationUnit): Transformer =
+    new PreTyperTransformer(unit)
+
+  class PreTyperTransformer(unit: CompilationUnit) extends Transformer {
+    override def transform(tree: Tree): Tree = tree match {
+      case tree: ClassDef if needsAnnotations(tree) =>
+        val newBody = tree.impl.body.map {
+          case vdef: ValDef if needsAnnotations(vdef) =>
+            treeCopy.ValDef(vdef, withWasPublic(vdef.mods), vdef.name,
+                vdef.tpt, vdef.rhs)
+
+          case ddef: DefDef if needsAnnotations(ddef) =>
+            treeCopy.DefDef(ddef, withWasPublic(ddef.mods), ddef.name,
+                ddef.tparams, ddef.vparamss, ddef.tpt, ddef.rhs)
+
+          case member => member
+        }
+        val newImpl =
+          treeCopy.Template(tree.impl, tree.impl.parents, tree.impl.self, newBody)
+        val newTree =
+          treeCopy.ClassDef(tree, tree.mods, tree.name, tree.tparams, newImpl)
+        super.transform(newTree)
+
+      case _ => super.transform(tree)
+    }
+  }
+
+  private def needsAnnotations(classDef: ClassDef): Boolean = {
+    classDef.name.toString == "$anon" &&
+    classDef.impl.body.exists {
+      case vdef: ValDef => needsAnnotations(vdef)
+      case ddef: DefDef => needsAnnotations(ddef)
+      case _ => false
+    }
+  }
+
+  private def needsAnnotations(vdef: ValDef): Boolean =
+    vdef.mods.isPublic
+
+  private def needsAnnotations(ddef: DefDef): Boolean =
+    ddef.mods.isPublic && ddef.name != nme.CONSTRUCTOR
+
+  private def withWasPublic(mods: Modifiers): Modifiers =
+    mods.withAnnotations(anonymousClassMethodWasPublicAnnotations)
+
+  private lazy val anonymousClassMethodWasPublicAnnotations: List[Tree] = {
+    val runtimePackage =
+      Select(Select(Ident(newTermName("scala")), newTermName("scalajs")), newTermName("runtime"))
+    val cls = Select(runtimePackage, newTypeName("WasPublicBeforeTyper"))
+    List(Apply(Select(New(cls), nme.CONSTRUCTOR), Nil))
```

Comment:
Create definitions for `newTermName("scalajs")` and `newTermName("WasPublicBeforeTyper")` (since this is going to become a def).